warehouse/manage/views.py

```diff
@@ -879,7 +879,10 @@ def create_macaroon(self):
 
         response = {**self.default_response}
         if form.validate():
-            macaroon_caveats = {"permissions": form.validated_scope, "version": 1}
+            # NOTE: The caveat order here is important: the UI assumes
+            # that the permissions caveat comes first in any recorded
+            # events.
+            macaroon_caveats = [{"permissions": form.validated_scope, "version": 1}]
```

Comment:
Highlighting this as another part I'm not satisfied with: we subsequently assume that `macaroon_caveats[0]` is the permissions caveat, e.g. when pretty-printing events. That's unlikely to change, but it's a piece of nasty implicit state here.

Reply:
(The other option is to embed `form.validated_scope` into the event instead of the entire `macaroon_caveats` list. That would avoid the implicit list position.)

Reply:
Or maybe we should update the UI to include all caveats?

Reply:
Alright, I've added a `caveat_detail` macro that handles both caveat types, removing the need for the implicit array positioning here.

The changes introduced with that macro don't currently manifest in the UI since we don't currently mint any API tokens with the new expiry caveat, but they'll show up once I begin the route work.